File: discord-bot/mcp-server/src/tools/inventory.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { prisma } from "../db.js";
import { getItemDisplayName, getItemCodesByTag, isVehicle } from "../utils/items.js";

export function registerInventoryTools(server: McpServer): void {
  // search_inventory - Search regiment inventory
  server.tool(
    "search_inventory",
    "Search regiment inventory for items by name, code, or slang (e.g., '12.7', 'mammon', 'bmat')",
    {
      regimentId: z.string().describe("Discord guild ID of the regiment"),
      query: z.string().optional().describe("Search term (item name, code, or slang)"),
      category: z.enum(["vehicles", "weapons", "ammo", "resources", "all"]).optional().describe("Filter by category"),
      limit: z.number().optional().default(20).describe("Max items to return"),
    },
    async ({ regimentId, query, category, limit }) => {
      // Get all stockpile items for this regiment
      const items = await prisma.stockpileItem.findMany({
        where: {
          stockpile: { regimentId },
        },
        select: {
          itemCode: true,
          quantity: true,
          crated: true,
          stockpileId: true,
        },
      });

      // Aggregate by itemCode
      const aggregateMap = new Map<string, {
        itemCode: string;
        displayName: string;
        totalQuantity: number;
        cratedQuantity: number;
        looseQuantity: number;
        stockpileIds: Set<string>;
      }>();

      for (const item of items) {
        const existing = aggregateMap.get(item.itemCode);
        const displayName = getItemDisplayName(item.itemCode);

        if (existing) {
          if (item.crated) {
            existing.cratedQuantity += item.quantity;
          } else {
            existing.looseQuantity += item.quantity;
          }
          existing.totalQuantity += item.quantity;
          existing.stockpileIds.add(item.stockpileId);
        } else {
          aggregateMap.set(item.itemCode, {
            itemCode: item.itemCode,
            displayName,
            totalQuantity: item.quantity,
            cratedQuantity: item.crated ? item.quantity : 0,
            looseQuantity: item.crated ? 0 : item.quantity,
            stockpileIds: new Set([item.stockpileId]),
          });
        }
      }

      // Convert to array
      let aggregated = Array.from(aggregateMap.values()).map(item => ({
        itemCode: item.itemCode,
        displayName: item.displayName,
        totalQuantity: item.totalQuantity,
        cratedQuantity: item.cratedQuantity,
        looseQuantity: item.looseQuantity,
        stockpileCount: item.stockpileIds.size,
        matchedTag: null as string | null,
      }));

      // Filter by category
      if (category === "vehicles") {
        aggregated = aggregated.filter(item => isVehicle(item.itemCode));
      }

      // Filter by search term
      if (query) {
        const searchLower = query.toLowerCase();
        const tagMatchedCodes = new Set(getItemCodesByTag(searchLower));

        aggregated = aggregated
          .filter(item =>
            item.displayName.toLowerCase().includes(searchLower) ||
            item.itemCode.toLowerCase().includes(searchLower) ||
            tagMatchedCodes.has(item.itemCode)
          )
          .map(item => ({
            ...item,
            matchedTag: tagMatchedCodes.has(item.itemCode) ? query.toUpperCase() : null,
          }));
      }

      // Sort by total quantity descending
      aggregated.sort((a, b) => b.totalQuantity - a.totalQuantity);

      // Apply limit
      aggregated = aggregated.slice(0, limit);

      return {
        content: [
          {
            type: "text" as const,
            text: JSON.stringify({
              query: query || null,
              category: category || "all",
              resultCount: aggregated.length,
              totalUniqueItems: aggregateMap.size,
              items: aggregated,
            }, null, 2),
          },
        ],
      };
    }
  );

  // get_item_locations - Find where specific item is stored
  server.tool(
    "get_item_locations",
    "Get list of stockpiles containing a specific item with quantities",
    {
      regimentId: z.string().describe("Discord guild ID of the regiment"),
      itemCode: z.string().describe("Item code to search for (e.g., 'RifleC', 'HEGrenade')"),
    },
    async ({ regimentId, itemCode }) => {
      // Get all stockpiles that have this item
      const stockpileItems = await prisma.stockpileItem.findMany({
        where: {
          itemCode,
          stockpile: { regimentId },
        },
        include: {
          stockpile: {
            select: {
              id: true,
              name: true,
              type: true,
              hex: true,
              locationName: true,
              updatedAt: true,
            },
          },
        },
        orderBy: { quantity: "desc" },
      });

      // Group by stockpile (separate entries for crated/loose)
      const stockpileMap = new Map<string, {
        id: string;
        name: string;
        type: string;
        hex: string;
        locationName: string;
        updatedAt: Date;
        looseQuantity: number;
        cratedQuantity: number;
        totalQuantity: number;
      }>();

      for (const item of stockpileItems) {
        const existing = stockpileMap.get(item.stockpile.id);

        if (existing) {
          if (item.crated) {
            existing.cratedQuantity += item.quantity;
          } else {
            existing.looseQuantity += item.quantity;
          }
          existing.totalQuantity += item.quantity;
        } else {
          stockpileMap.set(item.stockpile.id, {
            id: item.stockpile.id,
            name: item.stockpile.name,
            type: item.stockpile.type,
            hex: item.stockpile.hex,
            locationName: item.stockpile.locationName,
            updatedAt: item.stockpile.updatedAt,
            looseQuantity: item.crated ? 0 : item.quantity,
            cratedQuantity: item.crated ? item.quantity : 0,
            totalQuantity: item.quantity,
          });
        }
      }

      const stockpiles = Array.from(stockpileMap.values()).sort(
        (a, b) => b.totalQuantity - a.totalQuantity
      );

      // Calculate totals
      const totalQuantity = stockpiles.reduce((sum, s) => sum + s.totalQuantity, 0);
      const totalCrated = stockpiles.reduce((sum, s) => sum + s.cratedQuantity, 0);
      const totalLoose = stockpiles.reduce((sum, s) => sum + s.looseQuantity, 0);

      return {
        content: [
          {
            type: "text" as const,
            text: JSON.stringify({
              itemCode,
              displayName: getItemDisplayName(itemCode),
              totalQuantity,
              totalCrated,
              totalLoose,
              stockpileCount: stockpiles.length,
              stockpiles: stockpiles.map(s => ({
                ...s,
                location: `${s.hex} - ${s.locationName}`,
              })),
            }, null, 2),
          },
        ],
      };
    }
  );
}

File: mcp-server/src/tools/inventory.py
"""Inventory MCP tools - search and item locations."""

import json
from typing import TYPE_CHECKING, Any

from sqlalchemy import select

from ..db.session import get_session
from ..db.models import Stockpile, StockpileItem
from ..foxhole.items import get_item_display_name, get_item_codes_by_tag, is_vehicle

if TYPE_CHECKING:
    from ..server import McpServer


def register_inventory_tools(server: "McpServer") -> None:
    """Register inventory-related MCP tools."""

    async def search_inventory(args: dict[str, Any]) -> dict[str, Any]:
        """Search regiment inventory for items."""
        regiment_id = args["regimentId"]
        query = args.get("query")
        category = args.get("category")
        limit = int(args.get("limit", 20))

        async with get_session() as session:
            # Get all stockpile items for this regiment
            items_result = await session.execute(
                select(
                    StockpileItem.itemCode,
                    StockpileItem.quantity,
                    StockpileItem.crated,
                    StockpileItem.stockpileId,
                )
                .join(Stockpile, StockpileItem.stockpileId == Stockpile.id)
                .where(Stockpile.regimentId == regiment_id)
            )
            items = items_result.all()

        # Aggregate by itemCode
        aggregate_map: dict[str, dict[str, Any]] = {}

        for item in items:
            item_code = item.itemCode
            display_name = get_item_display_name(item_code)

            if item_code not in aggregate_map:
                aggregate_map[item_code] = {
                    "itemCode": item_code,
                    "displayName": display_name,
                    "totalQuantity": 0,
                    "cratedQuantity": 0,
                    "looseQuantity": 0,
                    "stockpileIds": set(),
                }

            entry = aggregate_map[item_code]
            if item.crated:
                entry["cratedQuantity"] += item.quantity
            else:
                entry["looseQuantity"] += item.quantity
            entry["totalQuantity"] += item.quantity
            entry["stockpileIds"].add(item.stockpileId)

        # Convert to list
        aggregated = [
            {
                "itemCode": v["itemCode"],
                "displayName": v["displayName"],
                "totalQuantity": v["totalQuantity"],
                "cratedQuantity": v["cratedQuantity"],
                "looseQuantity": v["looseQuantity"],
                "stockpileCount": len(v["stockpileIds"]),
                "matchedTag": None,
            }
            for v in aggregate_map.values()
        ]

        # Filter by category
        if category == "vehicles":
            aggregated = [item for item in aggregated if is_vehicle(item["itemCode"])]

        # Filter by search term
        if query:
            search_lower = query.lower()
            tag_matched_codes = set(get_item_codes_by_tag(search_lower))

            filtered = []
            for item in aggregated:
                if (
                    search_lower in item["displayName"].lower()
                    or search_lower in item["itemCode"].lower()
                    or item["itemCode"] in tag_matched_codes
                ):
                    if item["itemCode"] in tag_matched_codes:
                        item["matchedTag"] = query.upper()
                    filtered.append(item)
            aggregated = filtered

        # Sort by total quantity descending
        aggregated.sort(key=lambda x: x["totalQuantity"], reverse=True)

        # Apply limit
        aggregated = aggregated[:limit]

        return {
            "content": [
                {
                    "type": "text",
                    "text": json.dumps({
                        "query": query,
                        "category": category or "all",
                        "resultCount": len(aggregated),
                        "totalUniqueItems": len(aggregate_map),
                        "items": aggregated,
                    }, indent=2),
                },
            ],
        }

    server.tool(
        "search_inventory",
        "Search regiment inventory for items by name, code, or slang (e.g., '12.7', 'mammon', 'bmat')",
        {
            "regimentId": {
                "type": "string",
                "description": "Discord guild ID of the regiment",
                "required": True,
            },
            "query": {
                "type": "string",
                "description": "Search term (item name, code, or slang)",
            },
            "category": {
                "type": "string",
                "description": "Filter by category: vehicles, weapons, ammo, resources, all",
            },
            "limit": {
                "type": "integer",
                "description": "Max items to return",
                "default": 20,
            },
        },
        search_inventory,
    )

    async def get_item_locations(args: dict[str, Any]) -> dict[str, Any]:
        """Get list of stockpiles containing a specific item with quantities."""
        regiment_id = args["regimentId"]
        item_code = args["itemCode"]

        async with get_session() as session:
            # Get all stockpiles that have this item
            items_result = await session.execute(
                select(
                    StockpileItem.quantity,
                    StockpileItem.crated,
                    Stockpile.id,
                    Stockpile.name,
                    Stockpile.type,
                    Stockpile.hex,
                    Stockpile.locationName,
                    Stockpile.updatedAt,
                )
                .join(Stockpile, StockpileItem.stockpileId == Stockpile.id)
                .where(Stockpile.regimentId == regiment_id)
                .where(StockpileItem.itemCode == item_code)
                .order_by(StockpileItem.quantity.desc())
            )
            stockpile_items = items_result.all()

        # Group by stockpile (separate entries for crated/loose)
        stockpile_map: dict[str, dict[str, Any]] = {}

        for item in stockpile_items:
            stockpile_id = item.id

            if stockpile_id not in stockpile_map:
                stockpile_map[stockpile_id] = {
                    "id": stockpile_id,
                    "name": item.name,
                    "type": item.type.value if hasattr(item.type, 'value') else str(item.type),
                    "hex": item.hex,
                    "locationName": item.locationName,
                    "updatedAt": item.updatedAt.isoformat() if item.updatedAt else None,
                    "looseQuantity": 0,
                    "cratedQuantity": 0,
                    "totalQuantity": 0,
                }

            entry = stockpile_map[stockpile_id]
            if item.crated:
                entry["cratedQuantity"] += item.quantity
            else:
                entry["looseQuantity"] += item.quantity
            entry["totalQuantity"] += item.quantity

        stockpiles = sorted(
            stockpile_map.values(),
            key=lambda x: x["totalQuantity"],
            reverse=True,
        )

        # Calculate totals
        total_quantity = sum(s["totalQuantity"] for s in stockpiles)
        total_crated = sum(s["cratedQuantity"] for s in stockpiles)
        total_loose = sum(s["looseQuantity"] for s in stockpiles)

        return {
            "content": [
                {
                    "type": "text",
                    "text": json.dumps({
                        "itemCode": item_code,
                        "displayName": get_item_display_name(item_code),
                        "totalQuantity": total_quantity,
                        "totalCrated": total_crated,
                        "totalLoose": total_loose,
                        "stockpileCount": len(stockpiles),
                        "stockpiles": [
                            {
                                **s,
                                "location": f"{s['hex']} - {s['locationName']}",
                            }
                            for s in stockpiles
                        ],
                    }, indent=2),
                },
            ],
        }

    server.tool(
        "get_item_locations",
        "Get list of stockpiles containing a specific item with quantities",
        {
            "regimentId": {
                "type": "string",
                "description": "Discord guild ID of the regiment",
                "required": True,
            },
            "itemCode": {
                "type": "string",
                "description": "Item code to search for (e.g., 'RifleC', 'HEGrenade')",
                "required": True,
            },
        },
        get_item_locations,
    )
